vpr/src/route/rr_graph_fpga_interchange.cpp

```diff
@@ -0,0 +1,1706 @@
+
+            nodes.erase(key);
+
+            existing_nodes.emplace(key, root_node);
+
+            std::queue<intermediate_node*> builder;
+            builder.push(root_node);
+
+            while (!builder.empty()) {
+                intermediate_node* vertex;
+                vertex = builder.front();
+                builder.pop();
+                location loc = vertex->loc;
+                for (auto i : NODESIDES) {
+                    location other_loc = add_vec(loc, offsets[i]);
+                    bool temp = false;
+                    if (existing_nodes.find(other_loc) != existing_nodes.end()) {
+                        temp = true;
+                    } else if (nodes.find(other_loc) != nodes.end()) {
+                        intermediate_node* new_node = new intermediate_node(other_loc);
+                        temp = true;
+                        if (node_tile_to_segment_[std::make_tuple(node_id, loc_to_tile_[other_loc])] != -1)
+                            seg_id = node_tile_to_segment_[std::make_tuple(node_id, loc_to_tile_[other_loc])];
+                        nodes.erase(other_loc);
+                        existing_nodes.emplace(other_loc, new_node);
+                        builder.push(new_node);
+                    }
+                    vertex->links[i] = temp;
+                }
+            }
+            roots.insert(root_node);
+        } while (!nodes.empty());
+
+        for (auto& node : existing_nodes) {
+            if (pip_uses_node_loc(node_id, node.second->loc)) {
+                node.second->on_route = true;
+            }
+            if (pin_uses_node_loc(node_id, node.second->loc)) {
+                node.second->has_pins = true;
+                node.second->on_route = true;
+            }
+        }
+        return roots;
+    }
+
+    intermediate_node* update_end(intermediate_node* end, intermediate_node* node, enum node_sides side) {
+        intermediate_node* res = end;
+        int x, y;
+        x = end->loc.first > node->loc.first ? -1 : end->loc.first < node->loc.first ? 1 : 0;
+        y = end->loc.second > node->loc.second ? -1 : end->loc.second < node->loc.second ? 1 : 0;
+        switch (side) {
+            case LEFT_EDGE:
+                if (x < 0 || (x == 0 && y < 0))
+                    res = node;
+                break;
+            case TOP_EDGE:
+                if (y > 0 || (y == 0 && x > 0))
+                    res = node;
+                break;
+            case RIGHT_EDGE:
+                if (x > 0 || (x == 0 && y > 0))
+                    res = node;
+                break;
+            case BOTTOM_EDGE:
+                if (y < 0 || (y == 0 && x < 0))
+                    res = node;
+                break;
+            default:
+                VTR_ASSERT(false);
+                break;
+        }
+        return res;
+    }
+
+    void add_line(std::map<location, intermediate_node*>& existing_nodes,
```

Comment:
This function requires a high level comment

Reply:
Added in next commit